File: leetcode/top-interview-questions/Java/二叉搜索树中第K小的元素_ejg.java
class Solution {
    class Temp {
        int res = 0;
        int n = 0;
    }
    private int res = 0, n = 0;
    public int kthSmallest(TreeNode root, int k) {
        // dfs(root, n, k);
        // return res;
        Temp temp = new Temp();
        dfs(root, temp, k);
        return temp.res;
    }
    void dfs(TreeNode root, int n, int k) {
        // 错误写法，仅为了对比
        if (root.left != null) {
            dfs(root.left, n, k);
        }

        n++;
        System.out.println("cur n is: " + n);
        System.out.println("cur root.val is: " + root.val);
        
        if (n == k) {
            res = root.val;
            return ;
        }

        if (root.right != null) {
            dfs(root.right, n, k);
        }
    }

    void dfs(TreeNode root, Temp temp, int k) {
        if (root.left != null) {
            dfs(root.left, temp, k);
        }

        temp.n++;
        System.out.println("cur n is: " + temp.n);
        System.out.println("cur root.val is: " + root.val);
        if (temp.n == k) {
            temp.res = root.val;
            return ;
        }
        
        if (root.right != null) {
            dfs(root.right, temp, k);
        }
    }
}

File: leetcode/top-interview-questions/C++/二叉搜索树中第K小的元素_ejg.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int res = 0, n = 0;
public:
    int kthSmallest0(TreeNode* root, int k) {
        // 递归
        dfs(root, n, k);
        return res;
    }

    void dfs(TreeNode* root, int& n, const int k) {
        if (root->left != nullptr) {
            dfs(root->left, n, k);
        }
        
        n++;
        cout<<"cur n is: "<<n<<endl;
        cout<<"cur root->val is: "<<root->val<<endl;

        if (n == k) {
            res = root->val;
            return ;
        }
        
        if (root->right != nullptr) {
            dfs(root->right, n, k);
        }
    }

    void dfs1(TreeNode* root, int n, const int k) {
        // 这是错误写法，仅为了对比
        if (root->left != nullptr) {
            dfs(root->left, n, k);
        }

        n++;
        cout<<"cur n is: "<<n<<endl;
        cout<<"cur root->val is: "<<root->val<<endl;

        if (n == k) {
            res = root->val;
            return ; // 加了return 语句可以提前结束递归
        }
        
        if (root->right != nullptr) {
            dfs(root->right, n, k);
        }
    }

    int kthSmallest(TreeNode* root, int k) {
        //迭代
        stack<TreeNode*>sk;
        int res;
        int n = 1;
        TreeNode* cur = root;
        while(!sk.empty() || cur != nullptr){
            while(cur != nullptr){
                sk.push(cur);
                cur = cur->left;
            }
            cur = sk.top();
            sk.pop();
            if(n == k) return cur->val;
            n++;
            cur = cur->right;
        }
        return 0;
    }
};